pkg/querier/iterators/chunk_iterator.go

```diff
@@ -98,28 +104,15 @@ func (i *chunkIterator) AtFloatHistogram(fh *histogram.FloatHistogram) (int64, *
 
 	i.cachedTime = t
 	i.cachedFloatHistogram = fh
-	i.cachedFloatHistogramValid = true
-
+	i.cacheValid = true
 	return i.cachedTime, i.cachedFloatHistogram
 }
 
 func (i *chunkIterator) AtT() int64 {
-	if i.cachedValueValid || i.cachedHistogramValid || i.cachedFloatHistogramValid {
+	if i.cacheValid {
 		return i.cachedTime
 	}
-	switch i.valType {
-	case chunkenc.ValFloat:
-		t, _ := i.At()
-		return t
-	case chunkenc.ValHistogram:
-		t, _ := i.AtHistogram(nil)
-		return t
-	case chunkenc.ValFloatHistogram:
-		t, _ := i.AtFloatHistogram(nil)
-		return t
-	default:
-		panic(fmt.Errorf("chunkIterator: calling AtT with unknown chunk encoding %v", i.valType))
-	}
+	return i.it.Timestamp()
```

Comment:
I think we're on the right track , however this might lead to a degradation in context of the chunk merge iterator. Currently in the heap sort there we'd not do any call through an interface due to the caching. Please check if this iterator is used in that context and let's do some benchmark with the chunk merge iterator. https://github.com/grafana/mimir/blob/main/pkg/querier/iterators/chunk_merge_iterator.go

It might turn out that what we really want is to just cache the timestamp and literally nothing else - in case values are read once and never multiple times.

Reply:
I have gone through the implementation of `chunkMergeIterator`, and these are my observations:
- calls to `AtT()`, `AtFloat()`, `AtHistogram()`, `AtFloatHistogram()` never call a corresponding method on `chunkIterator`, and **always** return only cached values, if any.
- building a new `chunkMergeIterator`:
  - the [constructor](https://github.com/grafana/mimir/blob/ac8399a34fa6572d7c864244df14a4496492c921/pkg/querier/iterators/chunk_merge_iterator.go#L33) of `chunkMergeIterator` [creates a list of non-overlapping iterators](https://github.com/grafana/mimir/blob/ac8399a34fa6572d7c864244df14a4496492c921/pkg/querier/iterators/chunk_merge_iterator.go#L34), [calls a `Next()`](https://github.com/grafana/mimir/blob/ac8399a34fa6572d7c864244df14a4496492c921/pkg/querier/iterators/chunk_merge_iterator.go#L42) on each of the non-overlapping iterators, and finally [initializes a heap](https://github.com/grafana/mimir/blob/ac8399a34fa6572d7c864244df14a4496492c921/pkg/querier/iterators/chunk_merge_iterator.go#L58) of non-overlapping chunks.
  - [the `Next()` on a non-overlapping](https://github.com/grafana/mimir/blob/ac8399a34fa6572d7c864244df14a4496492c921/pkg/querier/iterators/chunk_merge_iterator.go#L276) iterator [calls the `Next()`](https://github.com/grafana/mimir/blob/ac8399a34fa6572d7c864244df14a4496492c921/pkg/querier/iterators/chunk_merge_iterator.go#L278) on the current chunk of type `chunkIterator` belonging to the former, which updates the timeseries and type of the underlying iterator of the latter.
  - the constructor doesn't cache any value, so calls to `AtT()`, `AtFloat()`, `AtHistogram()`, `AtFloatHistogram()` would return erroneous values, which is correct, since it is supposed to call `Next()` or `Seek()` before any of the `At*()` functions
- calling `chunkMergeIterator.Next()`: 
  - this method loads the value from the topmost element in the heap, by calling `At()`, `AtHistogram()` or `AtFloatHistogram()`, depending on the type, caches the returned timestamp and value. Then it calls the `Next()` on the topmpost element in the heap, so that its timestamp and value type are updated for a successive calls to `Next()`
  - this means that, after a call to `chunkMergeIterator.Next()`, it is safe to return cached values when calling `AtT()`, `At()`, `AtHistogram()` and `AtFloatHistogram()`, because these values have actually been loaded from the corresponding `chunkIterator` during the `chunkMergeIterator.Next()`.
- calling `chunkMergeIterator.Seek()`: 
  - this method calls the `Seek()` on all the non-overlapping iterators belonging to `chunkMergeIterator`, and then updates the heap.
  - after that, it loads the value from the topmost element in the heap, by calling `At()`, `AtHistogram()` or `AtFloatHistogram()`, depending on the type, caches the returned timestamp and value.
  - this means that, after a call to `chunkMergeIterator.Next()`, it is safe to return cached values when calling `AtT()`, `At()`, `AtHistogram()` and `AtFloatHistogram()`, because these values have actually been loaded from the corresponding `chunkIterator` during the `chunkMergeIterator.Seek()`.

Reply:
Discussed IRL: we'll cache the timestamp independently of the value via updates in Next() and Seek(). This will allow the `seriesIteratorHeap` to be extremely efficient. While at the same time get rid of calling AtHistogram() and AtFloatHistogram() with `nil`, which is for sure wasteful.